### PR TITLE
8268663: Crash when guards contain boolean expression

### PR DESCRIPTION
**Pull request by lahodaj**
When Flow analyzes boolean expressions, (un-)initialized are split into (un-)initialized when true and when false. When analyzing guarded patterns, we need to handle these split values. As the only way the case's body/statements can be executed is if the case's pattern matches, we need to use the initialized values when true.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268663](https://bugs.openjdk.java.net/browse/JDK-8268663): Crash when guards contain boolean expression

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/41/head:pull/41` \
`$ git checkout pull/41`

Update a local copy of the PR: \
`$ git checkout pull/41` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/41/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 41`

View PR using the GUI difftool: \
`$ git pr show -t 41`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/41.diff">https://git.openjdk.java.net/jdk17/pull/41.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/patterns/Guards.java [hunk 1]
```diff
@@ -84,6 +92,31 @@ String testBooleanSwitchExpression(Object o) {
         }
     }
 
+    String typeGuardIfTrueSwitchStatement(Object o) {
+        Object o2 = "";
+        switch (o) {
+            case Integer i && i == 0 && i < 1 && o2 instanceof String s: o = s + String.valueOf(i); return "true";
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/patterns/Guards.java:98]: could you please add tests with `||` in the condition?

Comment (lahodaj) [reply test/langtools/tools/javac/patterns/Guards.java:99]: Thanks, Vicente. I've added tests with `||` in https://github.com/openjdk/jdk17/pull/41/commits/d9e902b02b8a9916954fe4d2b665d4eaee3fb352 . One thing to note is that the tests are a bit different when writing something like: `case String s && s.isEmpty() || o2 instanceof Number n ->` `n` is not available here, as there is a path on which `o2` is not a `Number`.

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Pull request closed without merging**